thirdparty/patchmatchnet/models/net.py

```diff
@@ -0,0 +1,373 @@
+        self.conv4 = ConvBnReLU(16, 16, 3, 1, 1)
+        # [B,32,H/4,W/4]
+        self.conv5 = ConvBnReLU(16, 32, 5, 2, 2)
+        self.conv6 = ConvBnReLU(32, 32, 3, 1, 1)
+        self.conv7 = ConvBnReLU(32, 32, 3, 1, 1)
+        # [B,64,H/8,W/8]
+        self.conv8 = ConvBnReLU(32, 64, 5, 2, 2)
+        self.conv9 = ConvBnReLU(64, 64, 3, 1, 1)
+        self.conv10 = ConvBnReLU(64, 64, 3, 1, 1)
+
+        self.output1 = nn.Conv2d(64, 64, 1, bias=False)
+        self.inner1 = nn.Conv2d(32, 64, 1, bias=True)
+        self.inner2 = nn.Conv2d(16, 64, 1, bias=True)
+        self.output2 = nn.Conv2d(64, 32, 1, bias=False)
+        self.output3 = nn.Conv2d(64, 16, 1, bias=False)
+
+    def forward(self, x: torch.Tensor) -> Dict[str, torch.Tensor]:
+        """Forward method
+
+        Args:
+            x: images from a single view, in the shape of [B, C, H, W]. Generally, C=3
+
+        Returns:
+            output_feature: a python dictionary contains extracted features from stage_1 to stage_3
+                keys are "stage_1", "stage_2", and "stage_3"
+        """
+
+        output_feature = {}
+
+        conv1 = self.conv1(self.conv0(x))
+        conv4 = self.conv4(self.conv3(self.conv2(conv1)))
+
+        conv7 = self.conv7(self.conv6(self.conv5(conv4)))
+        conv10 = self.conv10(self.conv9(self.conv8(conv7)))
+
+        output_feature["stage_3"] = self.output1(conv10)
+
+        intra_feat = F.interpolate(conv10, scale_factor=2, mode="bilinear") + self.inner1(conv7)
+        del conv7, conv10
+        output_feature["stage_2"] = self.output2(intra_feat)
+
+        intra_feat = F.interpolate(intra_feat, scale_factor=2, mode="bilinear") + self.inner2(conv4)
+        del conv4
+        output_feature["stage_1"] = self.output3(intra_feat)
+
+        del intra_feat
+
+        return output_feature
+
+
+class Refinement(nn.Module):
+    """Depth map refinement network"""
+
+    def __init__(self):
+        """Initialize"""
+
+        super(Refinement, self).__init__()
+
+        # img: [B,3,H,W]
+        self.conv0 = ConvBnReLU(in_channels=3, out_channels=8)
+        # depth map:[B,1,H/2,W/2]
+        self.conv1 = ConvBnReLU(in_channels=1, out_channels=8)
+        self.conv2 = ConvBnReLU(in_channels=8, out_channels=8)
+        self.deconv = nn.ConvTranspose2d(
+            in_channels=8, out_channels=8, kernel_size=3, padding=1, output_padding=1, stride=2, bias=False
+        )
+
+        self.bn = nn.BatchNorm2d(8)
+        self.conv3 = ConvBnReLU(in_channels=16, out_channels=8)
+        self.res = nn.Conv2d(in_channels=8, out_channels=1, kernel_size=3, padding=1, bias=False)
+
+    def forward(
+        self, img: torch.Tensor, depth_0: torch.Tensor, depth_min: torch.Tensor, depth_max: torch.Tensor
+    ) -> torch.Tensor:
+        """Forward method
+
+        Args:
+            img: input source and reference images (B, 3, H, W)
+            depth_0: current depth map (B, 1, H//2, W//2)
+            depth_min: pre-defined minimum depth

```

Comment:
Things are looking good, @codyly. Would you mind adding what the shape of `depth_min` and `depth_max` are?